### DISCUSSION TITLE
Invoking a machine with existing state

### DISCUSSION THREAD

**Discussion by diestrin** [statelyai/xstate#1683] [Q&A]
Hi there

I'm developing a use case with xstate where I'm persisting the state in a db and hydrating that state back when invoked again. I have the case where a RequestMachine wants to invoke a previously created machine (stored with state in the DB).
I'm having troubles finding what's the right way to invoke this previously created machine with a running state as a service or actor of another machine, that gets created on every request in the backend.

I've looked a ton in the docs on how to spawn actors with a previous state, I know I can provide it with context, but not with state. I've also looked into using history nodes, but they don't persist the history in the context so I'm on the same place

Any help is appreciated

**Comment by JoshPolaris** [comment]
The documentation has a section on [persisting state](https://xstate.js.org/docs/guides/states.html#persisting-state), so my guess is that the same applies to invoking machines too.

```javascript
spawn(myMachine.resolveState(previousState));
```

**Reply by diestrin** [reply]
Spawn needs a Machine as an argument, resolveState returns a State object, I don't think this is the answer. I'm already using resolveState to re-hidrate the state of the machine when I recover it form the DB, it works. But spawn doesn't have a way to pass neither the interpreter of that machine, nor a way to define its initial state.

**Reply by kiliman** [reply]
Yes, `spawn()` uses a machine definition to create a new child service instance. It then calls `.start()` on this service.

We could extend `SpawnOptions` to include an optional `ResolvedState`. Then update the call to `.start()` in `spawnMachine()` to pass this state.

NOTE: some actors spawn additional child actors and store ActorRefs in context, so these child actors would also need to be re-spawned to get back to your original persisted state. We may need a hook somewhere to trigger this rehydration since spawn() should only be called during assignment.

Since actors typically have a consistent ID, it might be useful to include an optional state store so XState will automatically rehydrate actors based on ID.

**Reply by nprbst** [reply]
Could this be a reasonable way to add this to `v4`? #1898

**Comment by davidkpiano** [comment]
Use an event-sourcing approach: https://github.com/davidkpiano/xstate/discussions/1515

**Reply by diestrin** [reply]
This is not possible since my machines evoke side effects, if re-send all the actions again to the machines, they will execute those side-effects again.

To explain a little further:

I have 3 data types: A, B, C, they represent a physical object (an iot device).

I build the 3 different machines, so whenever I get a call in the backend I create an instance of this machine and store it in the DB deserializing it, so I will have hundreds of state objects recorded in the DB.
They really don't have a relationship on their own, one can execute different actions on those machines and it will all be good, with event sourcing I'm reading their changes and store it in DB again, as in the link you point me to.

Then I have a 4th machine type X, defining flows of actions/states that will be involved with these machines, this 4th flow gets created and destroyed for each request, since it manages the lifecycle of a single operation, no need to persist its state. But I want to interact from this temporal flow with the existing machines I have stored in DB.

Things I've think/tried:
- Looked for history states, my idea was to spawn from machine X a new machine of type A, B, or C using the DB context, and then point to the history state to go back to where it was
  - History is not stored in context, it's in the interpreter and thus I cannot do this
- I've think about having a service invoke a Promise, and the promise will load the machine from DB, start an interpreter, send the action and persist back the state, then complete the promise
  - Not a bad idea, but I'll loose the ability to pass in the context and to send multiples events back from the other machines, since this will be handled as a Promise invoke
- I might feel tempted to store the state object in the context, and then in the first status entry transition all the internal nodes to the appropriate state on the first load of the machine
  - Although I think this might work, I don't feel right storing the state in the context object, is that something I should try to avoid?

**Reply by Andarist** [reply]
I believe that this is something we are missing and it would be great to have an easy way to do this. Event-sourcing approach is now always feasible:
- for event-heavy systems it just doesnt scale well - its much cheaper to both store the required data for rehydration and actually rehydrate when dealing with a single snapshot of the “system” than when dealing with hundreds of events
- its currently not easy (i also dont really think that we need this) to record all events happening in the “system”. It’s only possible to track events sent externally to the system (by system i mean the root machine)

So I think that we should support rehydrating snapshota if the system in an easy way but it requires design and implementation. Probably wont happen any time soon - unless u maybe would like to help us with that 😉

**Reply by serapath** [reply]
any news? i have the same issue.

i assume right now the only way to do that is using option number 3 suggested by @diestrin, right?
so every machine would maybe get a default context object like `{ state: '', initial: '' }` and have `initial: 'init'` state, where the init satte would transition to one or maybe an array of states  given in the passed in context or otherwise default to whats written in initial. Of course that logic would include potential sub actors which have to undergo the same procedure.

does that make sense?

Until there is a native feature - what options exist to solve the issue mentioned by @diestrin ? :-)

**Reply by p-j** [reply]
There is a PR here: https://github.com/davidkpiano/xstate/pull/1903 by @justrhysism that started some work to address this. Maybe @serapath & @diestrin you would be interested in giving some feedback as to whether this addresses your respective issues?

**Reply by fe2bon4** [reply]
I've noticed that if all events have targets (events that also target the same state), following the [guide](https://xstate.js.org/docs/guides/states.html#persisting-state) actually started in the current state AND re-invoked the services in that state. 

However, if the last event sent to the interpreter does not have a target in the config, then the service just stops upon starting from the persisted state.

EDIT:  
It doesn't re-invoke the top level invocations. only the invocations within the targeted state of the event. 

Not sure if this is a bug or not. I'm using xstate version 4.20.0 and running this in a console only environment. 
It's a bit hacky but can probably work as a workaround for what you're trying to accomplish.

**Reply by Andarist** [reply]
@fe2bon4 could you prepare a codesandbox depicting what you are observing with a comment on what you have expected?

**Reply by fe2bon4** [reply]
Hi @Andarist 

I have an update. I dug in a bit more about restarting invoked actors with the clues that I've identified above. I figured out that you can restart the *invoked actors* the same way that you  restart *activities*. ([guide](https://xstate.js.org/docs/guides/activities.html#interpretation))


Though I also found that there was a line that is a bit misleading when restarting invoked services at Typescript would tell me that pushing the value from the the *state.activities* was incompatible with the type of items in *state.actions*.  I opted to ignore that and so it worked. 

here's my function that i use to re-hydrate the State object before passing it to an interpreter. 
It returns the resolved state with actions to restart the invoked actors.

```ts
import {
  interpret,
  State,
  StateMachine,
  actions,
  ActivityDefinition,
} from 'xstate';

const rehydrateState = (
  machine: StateMachine<any, any, any>,
  state_json: string = ''
): State<any, any> => {
  // Following from the guide so far.
  const state_definition =
    state_json === '' ? machine.initialState : JSON.parse(state_json);
  const machineState = State.create(state_definition);

  const resolvedState = machine.resolveState(machineState);

  Object.keys(resolvedState.activities)
    // Filter activities, and then add the start() action to the restored state
    .filter((activityKey) => {
      if (!resolvedState.activities[activityKey]) {
        return false;
      }

      const activityAction: ActivityDefinition<any, any> = resolvedState
        .activities[activityKey] as ActivityDefinition<any, any>;

      // Filter activities that have the same Id.
      const existingInvocation = resolvedState.actions.find((action) => {
        if (!action.activity) return false;
        return action.activity.id === activityAction.activity.id;
      });

      return !!!existingInvocation;
    })
    .forEach((activityKey) => {
      // From original guide in restarting activities.
      /* 
      Object.keys(restoredState.activities).forEach((activityKey) => {
        if (restoredState.activities[activityKey]) {
          // Filter activities, and then add the start() action to the restored state
          restoredState.actions.push(actions.start(activityKey));
        }
      });
      */

      // Actual Working Code that restarts the invoked actors.
      //@ts-ignore
      resolvedState.actions.push(resolvedState.activities[activityKey]);
    });

  // Log of what the current actions need to fire on start.
  console.log(resolvedState.actions);
  return resolvedState;
};
``` 


Not sure if a bug or not. I don't have much experience dealing with activities but our team mainly works on invoked actors and we needed a way to restart them. 

One of our use cases are for reconnecting to a Websocket server using an invoked callback as a service and getting events through that from the server. 

I have not tested my workaround to restart actual activities.
